File: fullstacks/nuxt/stores/plaza.ts
// ============================================================
// 广场状态：sort / filter / 关键词 + 分页 + 列表缓存
// ============================================================

import { defineStore } from "pinia";
import { ref } from "vue";
import { api } from "@/api/client";
import type {
  CapsuleListItem,
  Pagination,
  PlazaFilter,
  PlazaSort,
} from "@/types";

export const usePlazaStore = defineStore("plaza", () => {
  const sort = ref<PlazaSort>("new");
  const filter = ref<PlazaFilter>("all");
  const q = ref("");
  const page = ref(1);
  const pageSize = ref(15);

  const items = ref<CapsuleListItem[]>([]);
  const pagination = ref<Pagination | null>(null);
  const loading = ref(false);
  const error = ref<string | null>(null);

  // 请求序列号：fetch 内仅在自身是「最新发起」时才回写结果，避免快速切换排序时
  // 旧请求的响应覆盖新请求结果。用闭包变量而非 ref，避免触发不必要的订阅 re-render。
  let fetchSeq = 0;

  async function fetch() {
    const myId = ++fetchSeq;
    loading.value = true;
    error.value = null;
    try {
      const data = await api.plaza({
        sort: sort.value,
        filter: filter.value,
        q: q.value.trim() || undefined,
        page: page.value,
        pageSize: pageSize.value,
      });
      // 落后的请求直接丢弃
      if (myId !== fetchSeq) return;
      items.value = data.items;
      pagination.value = data.pagination;
    } catch (e) {
      if (myId !== fetchSeq) return;
      error.value = e instanceof Error ? e.message : "加载失败";
    } finally {
      // 仅最新请求负责切回 loading=false，避免被旧请求误关
      if (myId === fetchSeq) loading.value = false;
    }
  }

  function setSort(s: PlazaSort) {
    sort.value = s;
    page.value = 1;
    void fetch();
  }
  function setFilter(f: PlazaFilter) {
    filter.value = f;
    page.value = 1;
    void fetch();
  }
  function setQ(v: string) {
    q.value = v;
    page.value = 1;
    void fetch();
  }
  function setPage(p: number) {
    page.value = p;
    void fetch();
  }

  function patchFavorited(capsuleId: string, favorited: boolean, count: number) {
    items.value = items.value.map((it) =>
      it.id === capsuleId
        ? { ...it, favoritedByMe: favorited, favoriteCount: count }
        : it,
    );
  }

  return {
    sort,
    filter,
    q,
    page,
    pageSize,
    items,
    pagination,
    loading,
    error,
    fetch,
    setSort,
    setFilter,
    setQ,
    setPage,
    patchFavorited,
  };
});

File: desktop/flutter/lib/stores/plaza.dart
// 广场状态：sort/filter/q + 分页 + 列表缓存（= React stores/plaza.ts）。
// fetch 用自增序列号防并发乱序覆盖。
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../models/models.dart';
import 'providers.dart';

class PlazaState {
  final String sort; // hot | new
  final String filter; // all | opened | unopened
  final String q;
  final int page;
  final int pageSize;
  final List<CapsuleListItem> items;
  final Pagination? pagination;
  final bool loading;
  final String? error;

  const PlazaState({
    this.sort = 'new',
    this.filter = 'all',
    this.q = '',
    this.page = 1,
    this.pageSize = 15,
    this.items = const [],
    this.pagination,
    this.loading = false,
    this.error,
  });

  PlazaState copyWith({
    String? sort,
    String? filter,
    String? q,
    int? page,
    List<CapsuleListItem>? items,
    Pagination? pagination,
    bool? loading,
    String? error,
    bool clearError = false,
  }) =>
      PlazaState(
        sort: sort ?? this.sort,
        filter: filter ?? this.filter,
        q: q ?? this.q,
        page: page ?? this.page,
        pageSize: pageSize,
        items: items ?? this.items,
        pagination: pagination ?? this.pagination,
        loading: loading ?? this.loading,
        error: clearError ? null : (error ?? this.error),
      );
}

class PlazaNotifier extends Notifier<PlazaState> {
  int _seq = 0;

  @override
  PlazaState build() => const PlazaState();

  void setSort(String s) {
    state = state.copyWith(sort: s, page: 1);
    fetch();
  }

  void setFilter(String f) {
    state = state.copyWith(filter: f, page: 1);
    fetch();
  }

  void setQ(String q) {
    state = state.copyWith(q: q, page: 1);
    fetch();
  }

  void setPage(int p) {
    state = state.copyWith(page: p);
    fetch();
  }

  Future<void> fetch() async {
    final myId = ++_seq;
    state = state.copyWith(loading: true, clearError: true);
    try {
      final data = await ref.read(apiClientProvider).plaza(
            sort: state.sort,
            filter: state.filter,
            q: state.q.trim().isEmpty ? null : state.q.trim(),
            page: state.page,
            pageSize: state.pageSize,
          );
      if (myId != _seq) return;
      state = state.copyWith(items: data.items, pagination: data.pagination, loading: false);
    } catch (e) {
      if (myId != _seq) return;
      state = state.copyWith(loading: false, error: e.toString());
    }
  }

  void patchFavorited(String capsuleId, bool favorited, int count) {
    state = state.copyWith(
      items: [
        for (final it in state.items)
          if (it.id == capsuleId)
            CapsuleListItem(
              id: it.id,
              code: it.code,
              title: it.title,
              creator: it.creator,
              openAt: it.openAt,
              createdAt: it.createdAt,
              inPlaza: it.inPlaza,
              favoriteCount: count,
              isOpened: it.isOpened,
              favoritedByMe: favorited,
              favoritedAt: it.favoritedAt,
              contentPreview: it.contentPreview,
            )
          else
            it,
      ],
    );
  }
}

final plazaProvider = NotifierProvider<PlazaNotifier, PlazaState>(PlazaNotifier.new);
